lib/client.js

```diff
@@ -1278,6 +1280,120 @@ function _resume (client, sync) {
+
+  let finished = false
+  let bytesWritten = 0
+  function onFinished (err) {
+    if (finished) {
+      return
+    }
+
+    finished = true
+    socket[kWriting] = false
+    socket.removeListener('error', onFinished)
+
+    if (!err && contentLength !== null && bytesWritten !== contentLength) {
+      if (client[kStrictContentLength]) {
+        err = new RequestContentLengthMismatchError()
+      } else {
+        process.emitWarning(new RequestContentLengthMismatchError())
+      }
+    }
+
+    if (err) {
+      assert(client[kRunning] <= 1, 'pipeline should only contain this request')
+      util.destroy(socket, err)
+    }
+
+    if (socket.destroyed) {
+      return
+    }
+
+    if (bytesWritten === 0) {
+      if (expectsPayload) {
+        // https://tools.ietf.org/html/rfc7230#section-3.3.2
+        // A user agent SHOULD send a Content-Length in a request message when
+        // no Transfer-Encoding is sent and the request method defines a meaning
+        // for an enclosed payload body.
+
+        socket.write(`${header}content-length: 0\r\n\r\n\r\n`, 'ascii')
+      } else {
+        socket.write(`${header}\r\n`, 'ascii')
+      }
+    } else if (contentLength === null) {
+      socket.write('\r\n0\r\n\r\n', 'ascii')
+    }
+
+    if (socket[kParser].timeout && socket[kParser].timeoutType === TIMEOUT_HEADERS) {
+      // istanbul ignore else: only for jest
+      if (socket[kParser].timeout.refresh) {
+        socket[kParser].timeout.refresh()
+      }
+    }
+
+    resume(client)
+  }
+
+  let thrownError
+  try {
+    socket.on('error', onFinished)
+    for await (const chunk of body) {
+      if (finished || socket.destroyed) {
+        return
+      }
+
+      const len = Buffer.byteLength(chunk)
+      if (!len) {
+        continue
+      }
+
+      // TODO: What if not ended and bytesWritten === contentLength?
+      // We should defer writing chunks.
+      if (contentLength !== null && bytesWritten + len > contentLength) {
+        if (client[kStrictContentLength]) {
+          throw new RequestContentLengthMismatchError()
+        }
+
+        process.emitWarning(new RequestContentLengthMismatchError())
+      }
+
+      if (bytesWritten === 0) {
+        if (!expectsPayload) {
+          socket[kReset] = true
+        }
+
+        if (contentLength === null) {
+          socket.write(`${header}transfer-encoding: chunked\r\n`, 'ascii')
+        } else {
+          socket.write(`${header}content-length: ${contentLength}\r\n\r\n`, 'ascii')
+        }
+      }
+
+      if (contentLength === null) {
+        socket.write(`\r\n${len.toString(16)}\r\n`, 'ascii')
+      }
+
+      bytesWritten += len
+
+      if (socket.write(chunk)) {
+        continue
+      } else if (socket.destroyed) {
+        return
+      } else {
+        await EventEmitter.once(socket, 'drain')
+      }
+    }
+  } catch (err) {
+    thrownError = err
+  } finally {
+    onFinished(thrownError)
+  }
+}
```

Comment:
Look at the pump method in pipeline.js in node core

Reply:
Thanks, that really simplified moving everything to just one flow instead of having the `finished` variable. I didn't take the `AggregateError` part, as it's only in Node 16+, and I thought that it would be weird having type-guards and two different errors for two different Node versions.

Reply:
Note that I didn't add a `close` listener, which exists in `pump` but doesn't exist in the `stream` version. If a premature close error makes sense, I assume that it would make sense to add it in a following PR to both stream and iterator paths.

Reply:
Looks good!